### PR TITLE
Improve loss function handling in KerasClassifier

### PR DESCRIPTION
**Pull request by beat-buesser**
# Description

This pull request improves the handling of loss functions in `KerasClassifier` and enables the use of loss function generators which facilitates using models that predict logits. It updates the unittests for keras and tf.keras and tests each supported loss function type and implementation with probabilities and logits, where possible, and checks for identical results in `predict`, `loss_gradient` and `class_gradient`.

Fixes issues discovered as part of the discussion in #192

## Type of change

Please check all relevant options.

- [ ] Bug fix (non-breaking)
- [x] New feature (non-breaking)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: art/utils_test.py [hunk 1]
```diff
@@ -210,38 +211,109 @@ def get_classifier_kr(loss_name='categorical_crossentropy'):
...
+                    return keras.backend.sparse_categorical_crossentropy(y_true, y_pred, from_logits=True)
+
+                loss = sparse_categorical_crossentropy
+            else:
+                loss = keras.backend.sparse_categorical_crossentropy
     elif loss_name == 'kullback_leibler_divergence':
-        loss = keras.losses.kullback_leibler_divergence
-    elif loss_name == 'cosine_proximity':
-        loss = keras.losses.cosine_proximity
+        if loss_type == 'label':
```
Comment (ririnicolae) [inline art/utils_test.py]: Some of the if-elses around `loss_type` can probably be avoided by introducing a `dict` that maps the loss type to the action, then doing something like `loss = dict_loss_type[loss_type]`. The dict value could even be a function that does some of the checks required. This goes for all constructs of this type in the test models.

File: tests/classifiers/test_keras.py [hunk 2]
```diff
@@ -397,12 +397,171 @@ def test_repr(self):
...
+                        loss_type.format(loss_name, loss_type))
+
+            _run_tests(loss_name, loss_type, y_test_pred_expected, class_gradient_probabilities_expected,
+                       loss_gradient_expected, _from_logits=False)
+
+        # # ================= #
+        # # cosine_similarity #
+        # # ================= #
+        #
+        # loss_name = 'cosine_similarity'
```
Comment (ririnicolae) [inline tests/classifiers/test_keras.py]: Should this test stay commented?

File: tests/defences/test_spatial_smoothing.py [hunk 3]
```diff
@@ -39,6 +39,7 @@ def test_ones(self):
 
         # Start to test
         for window_size in range(1, 20):
+            print('Window size: {}'.format(window_size))
```
Comment (ririnicolae) [inline tests/defences/test_spatial_smoothing.py]: Why output the same info in logger and print?

File: tests/classifiers/test_keras.py [hunk 4]
```diff
@@ -397,12 +397,171 @@ def test_repr(self):
...
-            classifier.loss_gradient(x=self.x_test, y=self.y_test)
+
+        # prediction and class_gradient should be independent of logits/probabilities and of loss function
+
+        y_test_pred_expected = np.asarray([7, 1, 1, 4, 4, 1, 4, 4, 4, 4, 4, 4, 4, 4, 1, 4, 4, 7, 8, 4, 4, 4, 4, 3, 4, 4,
+                                           7, 4, 4, 4, 7, 4, 3, 4, 7, 0, 7, 7, 1, 1, 7, 7, 4, 0, 1, 4, 4, 4, 4, 4, 4, 7,
+                                           4, 4, 4, 4, 4, 1, 4, 4, 7, 4, 4, 4, 4, 4, 4, 4, 3, 0, 7, 4, 0, 1, 7, 4, 4, 7,
+                                           4, 4, 4, 4, 4, 7, 4, 4, 4, 4, 4, 1, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4])
+
+        class_gradient_probabilities_expected = np.asarray([0.0000000e+00, 0.0000000e+00, 0.0000000e+00, 2.3582461e-03,
```
Comment (ririnicolae) [inline tests/classifiers/test_keras.py]: The recent tendency has been to assert exact values in ART tests, but these are generally not very stable in the long run and require a lot of maintenance. Depending on the machine architecture and versions of packages, these numbers are not necessarily reproducible. E.g., in my case, versions of ART that pass on Travis fail locally. Should we think about another way of testing effectively?

File: tests/defences/test_spatial_smoothing.py (hunk 3 above)
Comment (beat-buesser) [reply tests/defences/test_spatial_smoothing.py]: I agree, the print statement needs to be removed again, it got forgotten. I had added it for debugging because this test sometimes fails and the test logging for this test is not showing up in the Travis logs.

Comment (beat-buesser) [reply tests/defences/test_spatial_smoothing.py]: Fixed in 614a65395b392fe959e50c267e6cc6517e4d22f5

File: tests/classifiers/test_keras.py (hunk 4 above)
Comment (beat-buesser) [reply tests/classifiers/test_keras.py]: I agree that recent versions of ART had too much variability in some of these new unittests, but I think we have removed or improved most of these unstable tests on `dev_1.1.0`. The unstable testes were mostly for classes containing algorithms with a lot of randomness and initially we thought that random seeds would fix it. I like this kind of tests in cases where the results are reproducible, because they run fast and capture changing behaviour quite well. But I'm also open to other types of testing too.

File: tests/classifiers/test_keras.py (hunk 2 above)
Comment (beat-buesser) [reply tests/classifiers/test_keras.py]: I will remove this loss function for now because its name changes between Keras/TF versions.

File: art/utils_test.py (hunk 1 above)
Comment (beat-buesser) [reply art/utils_test.py]: I agree and have tried to create a solution based on a dictionary but I feels like it becomes a quite complicated dictionary. I tried to improve the two functions by extending the docstrings and adding exception messages.

**Comment by beat-buesser** [comment]
Hi @ririnicolae Thank you for your great review! Ok, I'll wait with merging until tomorrow.
